src/context.rs

```diff
@@ -0,0 +1,87 @@
+use http::{Extensions, HeaderMap, Method, Uri, Version};
+use http_service::Body;
+use route_recognizer::Params;
+use std::sync::{Arc, Mutex};
+
+#[derive(Clone)]
+pub struct Context<AppData> {
+    app_data: Arc<AppData>,
+    context_data: Arc<ContextData>,
+}
+
+struct ContextData {
+    meta: http::request::Parts,
+    body: Mutex<http_service::Body>,
+    extensions: Mutex<Extensions>,
+    route_params: Params,
+}
+
+impl<AppData> Context<AppData> {
+    pub(crate) fn new(
+        app_data: Arc<AppData>,
+        req: http::Request<Body>,
+        route_params: Params,
+    ) -> Context<AppData> {
+        let (meta, body) = req.into_parts();
+        Context {
+            app_data,
+            context_data: Arc::new(ContextData {
+                meta,
+                body: Mutex::new(body),
+                extensions: Mutex::new(Extensions::new()),
+                route_params,
+            }),
+        }
+    }
+
+    pub fn method(&self) -> &Method {
+        &self.context_data.meta.method
+    }
+
+    pub fn uri(&self) -> &Uri {
+        &self.context_data.meta.uri
+    }
+
+    pub fn version(&self) -> Version {
+        self.context_data.meta.version
+    }
+
+    pub fn headers(&self) -> &HeaderMap {
+        &self.context_data.meta.headers
+    }
+
+    pub fn app_data(&self) -> &AppData {
+        &self.app_data
+    }
+
+    pub fn route_param(&self, key: &str) -> Option<&str> {
+        self.context_data.route_params.find(key)
+    }
+
+    pub async fn body_bytes(&mut self) -> std::io::Result<Vec<u8>> {
+        let body = self.replace_body(Body::empty());
+        await!(body.into_vec())
+    }
+
+    pub async fn body_string(&mut self) -> std::io::Result<String> {
+        let body_bytes = await!(self.body_bytes())?;
+        Ok(String::from_utf8(body_bytes).map_err(|_| std::io::ErrorKind::InvalidData)?)
+    }
+
+    pub async fn body_json<T: serde::de::DeserializeOwned>(&mut self) -> std::io::Result<T> {
+        let body_bytes = await!(self.body_bytes())?;
+        Ok(serde_json::from_slice(&body_bytes).map_err(|_| std::io::ErrorKind::InvalidData)?)
+    }
+
+    pub fn replace_body(&mut self, body: Body) -> Body {
+        std::mem::replace(&mut self.context_data.body.lock().unwrap(), body)
+    }
+
+    pub fn insert_extension<T: Send + Sync + 'static>(&mut self, ext: T) -> Option<T> {
+        self.context_data.extensions.lock().unwrap().insert(ext)
```

Comment:
*(worries about locking and unlocking mutexes often)*

Reply:
I think this should be okay performance-wise. This seems like a synchronization point that needs to happen, and we can always optimize later in some form or shape.

In any case, once we have some examples we can load test + analyze them, and quickly figure out which points form the bottleneck.